lib/brain.js

```diff
@@ -2762,9 +2787,92 @@ Brain.prototype.clearBill = function clearBill () {
 Brain.prototype.insertBill = function insertBill (bill) {
   console.assert(!this.bill || this.bill.fiat.eq(0), "bill fiat is positive, can't start tx")
   const cryptoCode = this.tx.cryptoCode
-  const exchangeRate = this.trader.rates(cryptoCode).cashIn
 
-  this.bill = Tx.createBill(bill, exchangeRate, this.tx)
+  // BACKWARDS_COMPATIBILITY 7.5.1
+  const serverVersion = this.trader.serverVersion
+  if (!serverVersion || semver.lt(serverVersion, '7.5.1-beta.0')) {
+    const exchangeRate = this.trader.rates(cryptoCode).cashIn
+    this.bill = Tx.createBillDeprecated(bill, exchangeRate, this.tx)
+  } else {
+    this.bill = Tx.createBill(bill, this.tx)
+  }
+}
+
+Brain.prototype._insertPromoCode = function _insertPromoCode () {
+  this._transitionState('insertPromoCode')
+
+  this._screenTimeout(() => {
+    this._cancelPromoCode()
+  }, this.config.promoCodeTimeout)
+}
+
+Brain.prototype._cancelPromoCode = function _cancelPromoCode () {
+  if (this.tx.direction === 'cashIn') this.returnToCashInState(false)
+  if (this.tx.direction === 'cashOut') this.returnToCashOutState(false)
+}
+
+Brain.prototype._submitPromoCode = function _submitPromoCode (data) {
+  const promoCode = data.input
+
+  this.trader.verifyCoupon(promoCode, this.tx).then(res => {
+    this.tx = Tx.update(this.tx, { discount: res.coupon.discount })
+
+    const rec = {
+      rates: {
+        rates: Tx.getRates(this.tx)[this.tx.cryptoCode],
+        cryptoCode: this.tx.cryptoCode,
+        coins: Tx.coins
+      },
+      credit: this._uiCredit()
+    }
+
+    this.browser().send(rec)
+
+    this.tx.direction === 'cashIn'
+      ? this.returnToCashInState(true)
+      : this.returnToCashOutState(true)
+  }).catch(err => {
+    console.log('Coupon code not found: Error ' + err)
+    this._transitionState('invalidPromoCode')
+    this._screenTimeout(() => {
+      this._cancelPromoCode()
+    }, this.config.promoCodeTimeout)
+  })
+}
+
+Brain.prototype.returnToScanState = function returnToScanState () {
+  const callback = this._start.bind(this)
+  this._screenTimeout(callback, this.config.qrTimeout)
+}
+
+Brain.prototype.returnToCashInState = function returnToCashInState (passTx) {
+  if (!this.tx.toAddress) return this.returnToScanState()
+  const returnState = this.tx.fiat.eq(0)
+    ? 'acceptingFirstBill'
+    : 'acceptingBills'
+
+  const callback = returnState === 'acceptingFirstBill'
+    ? this._idle.bind(this)
+    : this._sendCoins.bind(this)
+
+  passTx
+    ? this._transitionState(returnState, { tx: this.tx })
+    : this._transitionState(returnState)
```

Comment:
What happens if we pass the tx everytime?

Reply:
hmm, I just tested this and I seems that on both cashin and cashout, passing the transaction every time does not affect anything. Might have been some overengineering on my part, back when promo codes could be inserted after the first bill. Removing this and related code.